lib/ClangImporter/ClangImporter.cpp

```diff
@@ -4286,6 +4454,21 @@ ClangImporter::Implementation::loadNamedMembers(
   return Members;
 }
 
+void ClangImporter::Implementation::diagnoseMissingNamedMember(
+    const IterableDeclContext *IDC, DeclName name) {
+  Optional<clang::Module *> containingModule =
+      getClangSubmoduleForDecl(IDC->getDecl()->getClangDecl());
+  if (containingModule) {
```

Comment:
If there's no module we can just use the bridge header lookup table. This is: TODO.

Reply:
Ok thanks, I've never used dealt with bridging headers before but I'll take a look.

Reply:
It should be pretty trivial. Essentially when we use a bridge header instead of a module, there's a special lookup table that we can use instead of a lookup table associated with the clang module. So you can just say "if containingModule == nullptr then use Impl.BridgingHeaderLookupTable."

Reply:
Looks like I accidentally already handled this case? `findLookupTable` falls back to the `BridgingHeaderLookupTable` if there the passed ClangModule is null:
```cpp
SwiftLookupTable *ClangImporter::Implementation::findLookupTable(
                    const clang::Module *clangModule) {
  // If the Clang module is null, use the bridging header lookup table.
  if (!clangModule)
    return BridgingHeaderLookupTable.get();

  // Submodules share lookup tables with their parents.
  if (clangModule->isSubModule())
    return findLookupTable(clangModule->getTopLevelModule());

  // Look for a Clang module with this name.
  auto known = LookupTables.find(clangModule->Name);
  if (known == LookupTables.end()) return nullptr;

  return known->second.get();
}
```

Reply:
OK great. In that case this doesn't need to be an optional and you can remove the `if`.

Reply:
This also reminds me: please add a test where you use the bridging header. (It doesn't have to be a big test, just a sanity check.) Ping me if you are unsure how to do this.

Reply:
Seems to me the optional is still required. Null and none return value indicate different outcomes. Not entirely sure in what cases the decl does not have a valid associated module, but seems from a quick glance at the impementation this can happen if the decl is only forward declared?

```cpp
/// Returns the module \p D comes from, or \c None if \p D does not have
/// a valid associated module.
///
/// The returned module may be null (but not \c None) if \p D comes from
/// an imported header.
Optional<clang::Module *>
getClangSubmoduleForDecl(const clang::Decl *D,
                         bool allowForwardDeclaration = false);
```

```cpp
Optional<clang::Module *>
importer::getClangSubmoduleForDecl(const clang::Decl *D,
                                   bool allowForwardDeclaration) {
  const clang::Decl *actual = nullptr;

  // Put an Objective-C class into the module that contains the @interface
  // definition, not just some @class forward declaration.
  if (auto maybeDefinition = getDefinitionForClangTypeDecl(D)) {
    actual = maybeDefinition.getValue();
    if (!actual && !allowForwardDeclaration)
      return None;
  }

  if (!actual)
    actual = getFirstNonLocalDecl(D);

  return actual->getImportedOwningModule();
}
```

Reply:
I'll read a little closer, particularly about the forward declared case since we have diagnostics for it.

Reply:
I have handled this the same way loadNamedMembers does (assert the None case). Comments explains relatively well.
```cpp
TinyPtrVector<ValueDecl *>
ClangImporter::Implementation::loadNamedMembers(
    const IterableDeclContext *IDC, DeclBaseName N, uint64_t contextData) {

  auto *D = IDC->getDecl();
  auto *DC = D->getInnermostDeclContext();
  auto *CD = D->getClangDecl();
  auto *CDC = cast<clang::DeclContext>(CD);
  assert(CD && "loadNamedMembers on a Decl without a clangDecl");

  auto *nominal = DC->getSelfNominalTypeDecl();
  auto effectiveClangContext = getEffectiveClangContext(nominal);

  // There are 3 cases:
  //
  //  - The decl is from a bridging header, CMO is Some(nullptr)
  //    which denotes the __ObjC Swift module and its associated
  //    BridgingHeaderLookupTable.
  //
  //  - The decl is from a clang module, CMO is Some(M) for non-null
  //    M and we can use the table for that module.
  //
  //  - The decl is a forward declaration, CMO is None, which should
  //    never be the case if we got here (someone is asking for members).
  //
  // findLookupTable, below, handles the first two cases; we assert on the
  // third.

  auto CMO = getClangSubmoduleForDecl(CD);
  assert(CMO && "loadNamedMembers on a forward-declared Decl");

  auto table = findLookupTable(*CMO);
```

Reply:
Yeah, you should be able to use this:
```
  auto clangDecl = IDC->getDecl()->getClangDecl();
  auto *clangModule =
      getClangOwningModule(clangDecl, clangDecl->getASTContext());
  auto *lookupTable = ctx.getClangModuleLoader()->findLookupTable(clangModule); 
```

You can add an assertion if you want, but I don't think it's necessary.

Reply:
Applied this in combintation with the suggestion below.